File: milestone1/src/safety_node.cpp
#include "rclcpp/rclcpp.hpp"
/// CHECK: include needed ROS msg type headers and libraries
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include <cmath>


class Safety : public rclcpp::Node {
// The class that handles emergency braking

public:
    Safety() : Node("safety_node")
    {
        publisher = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(
            drive_topic,
            10
        );
        scan_subscription = this->create_subscription<sensor_msgs::msg::LaserScan>(
            lidarscan_topic,
            10,
            std::bind(&AutoDrive::scan_callback, this, std::placeholders::_1)
        );
        odom_subscription = this->create_subscription<nav_msgs::msg::Odometry>(
            odom_topic,
            10,
            std::bind(&AutoDrive::odom_callback, this, std::placeholders::_1)
        );
    }

private:
    double speed;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_publisher;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_subscription;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscription;

    void odom_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg) {
        // Update the current speed from the odometry message
        speed = msg->twist.twist.linear.x;
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) {
        const double threshold = 1.2; // Threshold for Time-to-Collision (TTC)
        auto ranges = scan_msg->ranges;
        const double angle_min = scan_msg->angle_min;
        const double angle_increment = scan_msg->angle_increment;

        for (size_t i = 0; i < ranges.size(); ++i) {
            double angle = angle_min + i * angle_increment;
            double v_long = speed * std::cos(angle);

            if (v_long > 0) {
                double iTTC = ranges[i] / v_long;
                if (iTTC < threshold) {
                    auto drive_msg = ackermann_msgs::msg::AckermannDriveStamped();
                    drive_msg.drive.speed = 0.0;
                    drive_publisher->publish(drive_msg);

                    RCLCPP_INFO(this->get_logger(), "EMERGENCY STOP: iTTC = %f", iTTC);
                    return;
                }
            }
        }
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Safety>());
    rclcpp::shutdown();
    return 0;
}

File: milestone1/scripts/safety_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

import numpy as np
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive

class SafetyNode(Node):
    """
    The class that handles emergency braking.
    """
    def __init__(self):
        super().__init__('safety_node')

        # One publisher should publish to the /drive topic with a AckermannDriveStamped drive message.
        self.publisher_ = self.create_publisher(AckermannDriveStamped, 'drive', 10)

        self.scan_subscription = self.create_subscription(
            LaserScan,
            'scan',
            self.scan_callback,
            10)
        self.odom_subscription = self.create_subscription(
            Odometry,
            'ego_racecar/odom',
            self.odom_callback,
            10)
        self.scan_subscription  # prevent unused variable warning
        self.odom_subscription

        self.speed = 0.

    def odom_callback(self, odom_msg):

        twist = odom_msg.twist.twist # http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/Twist.html
        v = twist.linear.x # Linear velocity, forward direction
        
        # Unused Odometry data
        # w = twist.angular # Vector3
        # pose = odom_msg.pose.pose
        # position = pose.position # Point
        # orientation = pose.orientation #Quaternion

        self.speed = v

        # Only for debugging purposes, not as helpful to us anymore now that things are working
        # prt = "v=" + str(v)
        # self.get_logger().info(prt)

    def scan_callback(self, scan_msg):
        
        threshold = 1.2 # Tuned this value with some trial and error, seems conservative enough
        # Yet still enables car to pass through narrow hallway in the back 
        ranges = scan_msg.ranges
        num = len(ranges)
        scan_time = scan_msg.scan_time 
        angle_min = scan_msg.angle_min
        angle_max = scan_msg.angle_max
        angle_increment = scan_msg.angle_increment
        v = self.speed

        for i in range(num):
            angle = angle_min + i*angle_increment # Compute angle for i'th element of ranges[]
            v_long = v * np.cos(angle) # Compute longitudinal v component i.e. in direction of this LiDAR beam
            if v_long <= 0:
                v_long = 0 # Compute {v_long}+ aka max(0,v_long), to capture risk of distance getting smaller
            else:
                iTTC = ranges[i] / v_long # Compute time to collision, using formula provided
            if v_long > 0 and iTTC < threshold: # Check if distance is decreasing, and doing so too fast
                new_msg = AckermannDriveStamped()
                new_msg.drive.speed = 0.0
                self.publisher_.publish(new_msg)
                prt = "EMERGENCY STOP: iTTC = " + str(iTTC)
                self.get_logger().info(prt)


def main(args=None):
    rclpy.init(args=args)
    safety_node = SafetyNode()
    rclpy.spin(safety_node)
    safety_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
